File: 300/316 - Remove Duplicate Letters.c
char* removeDuplicateLetters(char* s) {
    char *ret, c;
    int len, count[256] = {0}, visited[256] = {0}, i, r_len;
    
    if ((len = strlen(s)) == 0)
        return "";
    ret = (char *)malloc(len * sizeof(char));
    *ret = 'a' - 1;
    r_len = 1;
    
    for (i = 0; i < len; ++i)
        count[s[i]]++;
    
    for (i = 0; i < len; ++i) {
        c = s[i];
        count[c]--;
        
        if (visited[c])
            continue;
        
        while (c < ret[r_len - 1] && count[ret[r_len - 1]])
            visited[ret[--r_len]] = 0;

        ret[r_len++] = c;
        visited[c] = 1;
    }
    
    ret[r_len] = '\0';
    return ret + 1;
}

File: csharp/316.remove-duplicate-letters.cs
/*
 * @lc app=leetcode id=316 lang=csharp
 *
 * [316] Remove Duplicate Letters
 */

using System.Linq;

public class Solution {
    public string RemoveDuplicateLetters(string s) {
        int[] count = new int[26];
        bool[] visited = new bool[26];
        int[] str = s.Select(c => (int) (c - 'a')).ToArray();
        IList<int> ret = new List<int>();
        foreach (int c in str) {
            ++count[c];
        }
        foreach (int c in str) {
            --count[c];
            if (visited[c]) {
                continue;
            }
            while (ret.Count > 0) {
                int last = ret[ret.Count - 1];
                if (last < c || count[last] == 0) {
                    break;
                }
                ret.RemoveAt(ret.Count - 1);
                visited[last] = false;
            }
            ret.Add(c);
            visited[c] = true;
        }
        return new string(ret.Select(c => (char) (c + 'a')).ToArray());
    }
}

